include/optimization/cutting_plane.h

```diff
@@ -0,0 +1,1049 @@
+        return abs((exact - approx) / exact);
+    }
+
+
+    /**
+     * A class to save successive approximation values
+     */
+    class SlidingWindow {
+    public:
+        std::queue<double> evals;
+        int evalsSize; // how many values to store
+        int count;
+        double min;
+
+        SlidingWindow(int evalSize) {this->evalsSize = evalSize;count = 0;}
+
+        void push(double eval) {
+            if (count >= evalsSize) {
+                evals.pop();
+            }
+            else
+                count++;
+
+            evals.push(eval);
+            min = eval;
+        }
+
+        double getRelativeError() {
+            if (count < evalsSize)
+                return 1;
+
+            return abs((min - evals.front()) / min);
+        }
+
+        void half() {
+            for (int i=0 ; i<evals.size()/2 ; i++) {
+                evals.pop();
+                count--;
+            }
+        }
+    };
+
+
+    /**
+     * Compute the cutting plane c (x - point) <= 0  ==>  cx <= c point
+     * and add it as a new constraint in the existing polytope, in place of its last one, which will be redundant.
+     *
+     * @tparam Point class Point
+     * @tparam NT The numeric type
+     * @param c a vector holding the coefficients of the object function
+     * @param polytope An instance of Hpolytope
+     * @param point Where to create the cutting plane
+     */
+    template<class Point, typename NT>
+    void cutPolytope(VT &c, HPolytope<Point> &polytope, Point &point) {
+
+        unsigned int dim = polytope.dimension();
+
+        //add cx in last row of A
+        long j, i;
+
+        for (j = 0, i = polytope.num_of_hyperplanes() - 1 ; j < dim; j++) {
+            polytope.put_mat_coeff(i, j, c(j));
+        }
+
+        //add  < c,  point >  in last row of b
+        NT _b = c.dot(point.getCoefficients());
+        polytope.put_vec_coeff(polytope.num_of_hyperplanes() - 1, _b);
+    }
+
+
+    /**
+     * Store in dotProducts the dot product of c with all points in points
+     *
+     * @tparam Point class Point
+     * @tparam NT The numeric type
+     * @param c a vector holding the coefficients of the object function
+     * @param points a list of points
+     * @param dotProducts a list to store the dot products
+     */
+    template<class Point, typename NT>
+    void dotProducts(VT &c, std::list<Point> &points, std::list<NT> &dotProducts) {
+        for (auto p : points)
+            dotProducts.push_back(c.dot(p.getCoefficients()));
+    }
+
+
+    /**
+     * Find the point that minimizes the object function out of a list of points
+     *
+     * @tparam Point class Point
+     * @tparam NT The numeric type
+     * @param c a vector holding the coefficients of the object function
+     * @param randPoints a list of points
+     * @return point wich minimizes the object function
+     */
+    template<class Point, typename NT>
+    Point getMinimizingPoint(VT &c, std::list<Point> &randPoints) {
```

Comment:
`getMinimizingPoint` and `getPairMinimizingPoint` share a lot of functionality. Would you consider merging them?

Reply:
The former function returns the "best" point, while the latter the two "best" points. When I only need one, it saves a lot of computations (the list with points can be quite large).

Reply:
I mean to keep functionality of course but maybe merge the functions and have one function that given a flag returns the smallest or the two smallest elements. But it is up to you to decide.